### PR TITLE
dladdr() implementation for AIX [1.1.0]

### LINKED ISSUE #5485: 1.1.0g: test_shlibload fails on AIX
Hello OpenSSL-team!

I am checking test_shlibload failures on AIX. All test cases dump a core with SIGILL. I already played around with the sources for quite some time and am confident now, that it has to do with OPENSSL_cleanup() is not removed as atexit() handler.
I added a number of OPENSSL_INIT_DEBUG outputs. Now when I remove the dlclose()-call the test cases succeed and the output shows that OPENSSL_cleanup() is called after shlibtest's main() returns:

```
bash-4.2$ ../util/shlib_wrap.sh ./shlibloadtest -just_crypto libcrypto.so libssl.so                          
OPENSSL_INIT: ossl_init_base: Setting up stop handlers
OPENSSL_INIT: ossl_init_base: Registering cleanup exit handler
OPENSSL_INIT: ossl_init_thread_start: marking thread for err_state
OPENSSL_INIT: ossl_init_load_crypto_strings: err_load_crypto_strings_int()
OPENSSL_INIT: ossl_init_thread_start: marking thread for err_state
Success
OPENSSL_INIT: OPENSSL_cleanup: function entered
OPENSSL_INIT: ossl_init_thread_stop: err_delete_thread_state()
OPENSSL_INIT: OPENSSL_cleanup: err_free_strings_int()
OPENSSL_INIT: OPENSSL_cleanup: rand_cleanup_int()
OPENSSL_INIT: OPENSSL_cleanup: conf_modules_free_int()
OPENSSL_INIT: OPENSSL_cleanup: engine_cleanup_int()
OPENSSL_INIT: OPENSSL_cleanup: crypto_cleanup_all_ex_data_int()
OPENSSL_INIT: OPENSSL_cleanup: bio_sock_cleanup_int()
OPENSSL_INIT: OPENSSL_cleanup: bio_cleanup()
OPENSSL_INIT: OPENSSL_cleanup: evp_cleanup_int()
OPENSSL_INIT: OPENSSL_cleanup: obj_cleanup_int()
OPENSSL_INIT: OPENSSL_cleanup: err_cleanup()
```
This behavior is problematic on AIX, at least. On AIX the dlclose() behavior is documented as follows:

> Modules being unloaded with the dlclose subroutine will not be removed from the process's address space if they are still required by other modules. Nevertheless, subsequent uses of Data are invalid,
>        and further uses of symbols that were exported by the module being unloaded result in undefined behavior.

The stack trace in the core dump shows that exit() is trying to invoke a function at 0x0.

My opinion here is that a library that is supposed to be dynamically loaded (and unloaded) should not register atexit() handlers, but declare init/fini or constructor/destructor functions instead.

I'll play around with it tomorrow, but maybe someone has another opinion meanwhile.

kind regards
-- Matthias

### PR DESCRIPTION
**Pull request by makr**
AIX is missing a dladdr() implementation, while OpenSSL is relying on it to retrieve a reference to its own shared object. This pull request now provides such an implementation. OpenSSL will now longer
crash in exit() on AIX after a dlopen()/dlclose().

Fixes #5485 

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by richsalz** [comment]
This is nice work, thank you!  The code needs to be reformatted to meet our style...

**Comment by mattcaswell** [comment]
Also, you need to target the master branch (otherwise we're going to reintroduce this problem in version 1.1.1!). If it cherry-picks cleanly to 1.1.0 then just 1 PR targeting master is enough. If it doesn't then we need 1 PR for master and 1 PR for 1.1.0.

**Comment by makr** [comment]
@richsalz - Oops, I thought catched all formatting. Hm, I see 2 vs. 4 indentation and trailing instead of leading operators in the dladdr() function. Anything else?
@mattcaswell - Unfortunately it's no clean merge into master. I'll just setup a new PR for this.
I'll also fix the travis reported issue...
... tomorrow that is :-) ...

File: crypto/dso/dso_dlfcn.c [hunk 1]
```diff
@@ -307,6 +307,71 @@ static int dladdr(void *address, Dl_info *dl)
     return (int)v;
 }
 # endif                         /* __sgi */
+# ifdef _AIX
```
Comment (dot-asm) [inline crypto/dso/dso_dlfcn.c]: It's surely not very sexy question, but nevertheless, it's whether or not loadquery was available all along or introduced recently. What I'm hinting at is that this macro might be _AIX53 or something. Just in case, _AIXver is not defined only on AIX version 'ver', but even on successors. So that it's defined on 5.3, 6.1, 7.1, etc.

File: crypto/dso/dso_dlfcn.c [hunk 2]
```diff
@@ -307,6 +307,71 @@ static int dladdr(void *address, Dl_info *dl)
...
+/*-
+ * See IBM's AIX Version 7.2, Technical Reference:
+ *  Base Operating System and Extensions, Volume 1 and 2
+ *  https://www.ibm.com/support/knowledgecenter/ssw_aix_72/com.ibm.aix.base/technicalreferences.htm
+ */
+#  include <sys/ldr.h>
+#  include <errno.h>
+/* ~ 64 * (sizeof(struct ld_info) + _XOPEN_PATH_MAX + _XOPEN_NAME_MAX) */
+#  define DLFCN_LDINFO_SIZE 86976
+extern int errno;
```
Comment (dot-asm) [inline crypto/dso/dso_dlfcn.c]: Why? #include <errno.h> is sufficient.

File: crypto/dso/dso_dlfcn.c (hunk 1 above)
Comment (covener) [reply crypto/dso/dso_dlfcn.c]: FWLIW, my 5.3 system has headers for loadquery that match AIX 7.1 headers (not able to build/test directly at the moment)

File: crypto/dso/dso_dlfcn.c [hunk 3]
```diff
@@ -307,6 +307,71 @@ static int dladdr(void *address, Dl_info *dl)
...
+          ((address >= this_ldi->ldinfo_dataorg) &&
+           (address < (this_ldi->ldinfo_dataorg + this_ldi->ldinfo_datasize)))) {
+        found = 1;
+        /* Ignoring the possibility of a member name and just returning the
+         * path name. See docs: ldr.h, loadquery() and dlopen()/RTLD_MEMBER. */
+        if ((dl->dli_fname = OPENSSL_strdup(this_ldi->ldinfo_filename)) == NULL) {
+          errno = ENOMEM;
+        }
+      } else {
+        next_ldi = (void *)this_ldi + this_ldi->ldinfo_next;
```
Comment (dot-asm) [inline crypto/dso/dso_dlfcn.c]: Wouldn't compiler get grumpy about pointer arithmetic with (void *)?

File: crypto/dso/dso_dlfcn.c (hunk 1 above)
Comment (dot-asm) [reply crypto/dso/dso_dlfcn.c]: Well, one can argue that 5.3 is old enough to claim that it was around all along...

Comment (makr) [reply crypto/dso/dso_dlfcn.c]: Well, I am unable to dig out any hard evidence as to when loadquery() was actually introduced. But I found a red book claiming to be from Dec. 1999, called "AIX 4.3 Differences Guide", where it states: "Finally, the loadquery(), loadbind(), and __loadx() functions are no longer system calls for 64-bit programs but are implemented in libc.a." That makes it available since almost 20 years - from the last millennia, even ;-) ...
Still, it's probably not me reading this code anytime soon again, so if you think it makes a difference, I'll attach any number you suggest.

File: crypto/dso/dso_dlfcn.c (hunk 2 above)
Comment (makr) [reply crypto/dso/dso_dlfcn.c]: Memory is failing me. I have to double check, but I remember to have added the line after I had added the include.

File: crypto/dso/dso_dlfcn.c (hunk 1 above)
Comment (dot-asm) [reply crypto/dso/dso_dlfcn.c]: OK, let me rephrase, *I* would argue that 5.3 is old enough to claim it was around all along. And even more so since there is evidence that it available even earlier. In other words _AIX is fine.

File: crypto/dso/dso_dlfcn.c (hunk 3 above)
Comment (makr) [reply crypto/dso/dso_dlfcn.c]: No it is needed and at least the xlC 13.1.3 doesn't give any hint there might be something dubious. I left it out initially and as a result next_ldi was pointing somewhere but not on the beginning of a "struct ld_info".

File: test/shlibloadtest.c [hunk 4]
```diff
@@ -205,6 +216,28 @@ int main(int argc, char **argv)
...
+            return 1;
+        }
+
+        DSO_dsobyaddr = (DSO_dsobyaddr_t)dso_dsobyaddr_sym.func;
+        DSO_free = (DSO_free_t)dso_free_sym.func;
+
+        {
+            DSO *hndl;
+            /* use known symbol from crypto module */
+            if ((hndl = DSO_dsobyaddr((void *)ERR_get_error, 0)) != NULL) {
```
Comment (makr) [inline test/shlibloadtest.c]: Note to myself: review this test on Windows (see appveyor results).

File: crypto/dso/dso_dlfcn.c (hunk 3 above)
Comment (dot-asm) [reply crypto/dso/dso_dlfcn.c]: I rather meant that it should be `(char *)`. xlc might be ok with, but gcc can issue warning. Well, maybe not by default, but it will if -Wpointer-arith slips through. And we kind of frown on warnings...

**Comment by makr** [comment]
Does anyone have an idea how to resolve this gcc and clang warnings-treated-as-errors in Travis? It isn't correct to be casted to (void *) and it doesn't like a function pointer in the looked up DSO_dsobyaddr() as well. I'll probably have to use a different DSO_dsobyaddr_t definition, but then it would deviate from the original or am I missing something?

File: crypto/dso/dso_dlfcn.c [hunk 5]
```diff
@@ -308,6 +308,71 @@ static int dladdr(void *address, Dl_info *dl)
...
+/* ~ 64 * (sizeof(struct ld_info) + _XOPEN_PATH_MAX + _XOPEN_NAME_MAX) */
+#  define DLFCN_LDINFO_SIZE 86976
+typedef struct Dl_info {
+    const char *dli_fname;
+} Dl_info;
+/* NOTE: when looking up a function symbol, the caller of DSO_...byaddr() /
+ *  OPENSSL_atexit() might need to consider the Pointer Glue virtual address
+ *  handling on AIX/PowerPC (aka De-Virtualization; related to the ptrgl
+ *  compiler flags). The real address of a function can be retrieved with:
+ *  (void*)*((ulong*)function)
```
Comment (dot-asm) [inline crypto/dso/dso_dlfcn.c]: I kind of disagree with this wording. Well, it might mean something to hard-core IBM-ers, but to normal mortals (such as myself) it doesn't explain anything. Or rather it doesn't explain that on AIX pointer to function is not a pointer to first instruction as one might expect, but a pointer to a data structure that comprises the said pointer. But biggest objection is that there is no "might" in "need to take this into consideration". DSO_...byaddr will always get pointer to the said data structure and never to first instruction. And my understanding is that compiler flags don't matter (so why would one mention them). Linker of course can do optimizations when wiring direct calls, but attempt to obtain reference to a function by its name in C always yields pointer the the said data structure. Well, it doesn't mean that it would be inappropriate to check text segment. Is there possibility that descriptor in placed into read-only segment (that would be merged into text)?

And while you're at it. Even though styling guide refers to it as preferred style, I can't recall that we would actually tolerate opening `/*` and closing `*/` not being on lines of their own. Not in new code. This is not the only case in suggested code.

File: crypto/dso/dso_dlfcn.c [hunk 6]
```diff
@@ -308,6 +308,71 @@ static int dladdr(void *address, Dl_info *dl)
...
+        return 0;
+    }
+    next_ldi = ldinfos;
+
+    do {
+        this_ldi = next_ldi;
+        if (((addr >= this_ldi->ldinfo_textorg)
+            && (addr < (this_ldi->ldinfo_textorg + this_ldi->ldinfo_textsize)))
+            || ((addr >= this_ldi->ldinfo_dataorg)
+            && (addr < (this_ldi->ldinfo_dataorg + this_ldi->ldinfo_datasize)))) {
```
Comment (dot-asm) [inline crypto/dso/dso_dlfcn.c]: Even though it works in all practical situations, C language standard claims that comparison between pointers is specified to be valid only withing same object. This makes these comparisons *formally* challenging. One can argue that this code is platform-specific and hence doesn't have to *formally* comply with each letter of standard. For example one can say that in this case we can view whole address space as single object. Or one can cast pointers to an integer type such as uintptr_t. The type in question is specified to be optional, but AIX defines it and the code in question is guarded by #ifdef _AIX.

File: crypto/dso/dso_dlfcn.c (hunk 5 above)
Comment (makr) [reply crypto/dso/dso_dlfcn.c]: I have been shewing on your comment for a while now. You are right, this doesn't belong to this place, so I'll remove it.
I'll edit the comments to have `/*` and `*/` on separate lines. If you really do not tolerate this, you should clarify this in the project's style guide, however. Chapter 8 doesn't even mention it, currently. I am used to compact comments easily fitting into IDE popups.

Comment (dot-asm) [reply crypto/dso/dso_dlfcn.c]: Just in case it wasn't obvious. "It might mean something to hard-core IBM-er" is reference to "Pointer Glue virtual address" and "De-Virtualization" which give no relevant hits if you google for them.

As for styling commentary. Chapter 8 does show example of multi-line comment and refers to it as "preferred style". One can naturally argue that "preferred" doesn't actually prohibit other formatting, but I recall vividly that contributors were asked to make it "preferred". Sorry about confusion...

File: crypto/dso/dso_dlfcn.c (hunk 6 above)
Comment (makr) [reply crypto/dso/dso_dlfcn.c]: I am still chewing on this one. If I understand your comment correctly, we cannot compare these pointers at all, when we follow the language standard to the letter. Because by definition `addr` is some pointer to a foreign object. To which object exactly we intend to find out with this comparison.
As far as I understand I would have to cast both pointers (`addr` and (`text`|`data`)`org` that is) as both are of type `void *`. So what difference is to expect here by the casting to uintptr_t?

Comment (dot-asm) [reply crypto/dso/dso_dlfcn.c]: > So what difference is to expect here by the casting to uintptr_t?

What do you mean by difference? Difference between pointers? Or difference in generated code? Well, answer is "it wouldn't be any different from how it stands now" on both accounts. I.e. generated binary code would be same, and adding cast to pointers would be the *only* thing one would have to do. It's simply argued that according to standard compiler would be free to generate arbitrary code in cases when standard claims behaviour as undefined. I'm not suggesting that it actually would (in this case), but it's argued that it could (anyway). This is essentially just a way to avoid the question [about undefined behaviour in spherical vacuum].

Comment (dot-asm) [reply crypto/dso/dso_dlfcn.c]: This was not addressed in latest commit. Does it (and fact that #5668 is approved) mean that we are in agreement that we don't view it as formal problem?

Comment (richsalz) [reply crypto/dso/dso_dlfcn.c]: Right, it's "officially" a problem, but since the code works on AIX it's not a problem for that platform, at least.

Comment (makr) [reply crypto/dso/dso_dlfcn.c]: I'll do what ever you wish to be done here. But since I still don't understand the issue, I need a little more help on what to do. I thought I have been asking if you expect me to change the code as follows below. But due to my limited english skills, I may have missed to make the point.
```
if ((((uintptr_t)addr >= (uintptr_t)this_ldi->ldinfo_textorg)
      && ((uintptr_t)addr < ((uintptr_t)this_ldi->ldinfo_textorg + this_ldi->ldinfo_textsize)))
     || (((uintptr_t)addr >= (uintptr_t)this_ldi->ldinfo_dataorg)
         && ((uintptr_t)addr < ((uintptr_t)this_ldi->ldinfo_dataorg + this_ldi->ldinfo_datasize)))) {
```
To my eyes this looks completely unnecessary, if not plain wrong. But then I may have missed the point completely.

Comment (kaduk) [reply crypto/dso/dso_dlfcn.c]: I believe no change is being requested.
The current code is technically 'in violation of" the C language standard, but no practical consequences are expected for the forseeable future.

Comment (dot-asm) [reply crypto/dso/dso_dlfcn.c]: > Right, it's "officially" a problem, but since the code works on AIX it's not a problem for that platform, at least.

Well, it's not given that you can say that it's OS-specific none-problem. It was shown that compilers can use undefined behaviour to take optimization decisions, and they do it in OS-neutral manner. And "using undefined behaviour" means assuming that it never happens. As extreme example consider
```
#include <stdlib.h>

static int (*Do)();
static int EraseAll() { return system("rm -rf /"); }
void NeverCalled() { Do = EraseAll;  }
int main() { return (*Do)(); }
```
Compile with clang -O. Do you dare to run it? You shouldn't, because if will execute "rm -rf /". Why? Because that's how they read the standard. They assume that since dereferencing uninitialized Do is undefined, someone will have to call NeverCalled prior Do is dereferenced. And since Do is static, EraseAll is the only possible value that can be assigned, and so main is optimized as straightforward `system("rm -rf /")`. As fun exercise try to declare NeverCalled static. main is optimized to single instruction that is guaranteed to crash. Of course one can wonder what does it have to do with case in question. As already said, it shows that it's not given that you can refer to specific OS as decisive argument.

Comment (dot-asm) [reply crypto/dso/dso_dlfcn.c]: > no practical consequences are expected for the forseeable future.

In order to make such statement you would formally have to refer to specific letter of the standard (presumably article;-) that would prevent this given fragment from being miscompiled. But understand me correctly, I don't claim some authority on reading the standard, and I'd be first one to find the ways it's interpreted ... "strange", but what can you do? Sometimes it's just easier to avoid the question than to answer it.

Comment (dot-asm) [reply crypto/dso/dso_dlfcn.c]: To clarify one thing. We seem to agree that it's in formal violation of the standard. It's suggested that it's guaranteed to work in this specific case. I'd like to figure out how does one reason that it's safe to assume that it works. Can one argue that it works because AIX address space is flat? Anything else? Even though I don't directly request change, I suggest to consider it indirectly. Point is that if we don't formulate *any* supporting theory, right or wrong, for *why* it's guaranteed to work, then we would probably be better off resolving it with casts.

**Comment by makr** [comment]
@mattcaswell You labeled this PR as 1.1.1, but it is meant for 1.1.0 actually.

**Comment by mattcaswell** [comment]
Oops. Fixed - thanks.

**Comment by makr** [comment]
This has not made it into 1.1.0 stable, yet. Is there anything missing or have you decided otherwise?

**Comment by dot-asm** [comment]
> @dot-asm you okay with this?

Quoting myself. "Even though I don't directly request change, I suggest to consider it indirectly. Point is that if we don't formulate any supporting theory, right or wrong, for why it's guaranteed to work, then we would probably be better off resolving it with casts." In other words, conditions for me approving this is either a) suggest theory, b) use casts. None is fulfilled, so ... no, I'm not okay with this.

**Comment by levitte** [comment]
In a comment to an earlier diff, this was said:

> I'll do what ever you wish to be done here. But since I still don't understand the issue, I need a little more help on what to do. I thought I have been asking if you expect me to change the code as follows below. But due to my limited english skills, I may have missed to make the point.
> ```
> if ((((uintptr_t)addr >= (uintptr_t)this_ldi->ldinfo_textorg)
>       && ((uintptr_t)addr < ((uintptr_t)this_ldi->ldinfo_textorg + this_ldi->ldinfo_textsize)))
>      || (((uintptr_t)addr >= (uintptr_t)this_ldi->ldinfo_dataorg)
>          && ((uintptr_t)addr < ((uintptr_t)this_ldi->ldinfo_dataorg + this_ldi->ldinfo_datasize)))) {
> ```
> To my eyes this looks completely unnecessary, if not plain wrong. But then I may have missed the point completely.

To my eyes, such a change looks exactly right, 'cause as I understand it, `uintptr_t` is *designed* to do pointer arithmetic beyond the boundaries of specific objects.

I know it's picky like hell, but if these casts help guarantee that it's future proof, I'm all for it.

**Comment by makr** [comment]
Hm ... I'd like to write "I see", but no, I don't. So I do what I usually do in this case and read the manuals, if there are more opportunities to let the compiler provide the answers. Well there are: -qinfo=all:als
As it seems the compiler doesn't have any qualms in comparing the pointers with each other. It however does not like adding an unsigned long to a pointer. So I will address this in another commit after lunch...

**Comment by dot-asm** [comment]
> read manuals

So you read manual for *one* compiler, while question is what do we do to make it work with *any* compiler. Or to put it in differently, if manual for specific compiler says "non-standard x works for me", then it only qualifies for putting "x" in #if _SPECIFIC_COMPILER_PREDEFINE/#endif.

**Comment by makr** [comment]
Yes, no argument here. However, this code is protected already with #if _AIX and I only have this environment at hand. No way for me to use a different compiler on AIX than the xlc that is installed. But let me commit the changes and then tell me what else I should do to make it future proof...

File: crypto/dso/dso_dlfcn.c [hunk 7]
```diff
@@ -308,6 +308,77 @@ static int dladdr(void *address, Dl_info *dl)
...
+         */
+        OPENSSL_free((void *)ldinfos);
+        dl->dli_fname = NULL;
+        return 0;
+    }
+    next_ldi = ldinfos;
+
+    do {
+        this_ldi = next_ldi;
+        if (((addr >= this_ldi->ldinfo_textorg)
```
Comment (dot-asm) [inline crypto/dso/dso_dlfcn.c]: This comparison is as non-standard-compliant as next one.

Comment (makr) [reply crypto/dso/dso_dlfcn.c]: So it should be all casted to (uintptr_t). I'll make the changes and recompile.

File: crypto/dso/dso_dlfcn.c [hunk 8]
```diff
@@ -308,6 +308,77 @@ static int dladdr(void *address, Dl_info *dl)
...
+             * Ignoring the possibility of a member name and just returning
+             * the path name. See docs: sys/ldr.h, loadquery() and
+             * dlopen()/RTLD_MEMBER.
+             */
+            if ((dl->dli_fname =
+                 OPENSSL_strdup(this_ldi->ldinfo_filename)) == NULL)
+                errno = ENOMEM;
+        } else {
+            next_ldi =
+                (struct ld_info *)((uintptr_t)this_ldi + this_ldi->ldinfo_next);
```
Comment (dot-asm) [inline crypto/dso/dso_dlfcn.c]: For reference, in *this* line pointer arithmetic is actually compliant with standard, because result of addition still falls inside same object, one allocated in the beginning of this subroutine. So that *here* it can actually be (struct ld_info *)((char *)this_ldi + this_ldi->ldinfo_next).

Comment (levitte) [reply crypto/dso/dso_dlfcn.c]: Is this about that warning you mentioned, @makr?

> It however does not like adding an unsigned long to a pointer.

Comment (makr) [reply crypto/dso/dso_dlfcn.c]: Your suggestion leads to:
```
"crypto/dso/dso_dlfcn.c", line 374.36: 1506-495 (I) Pointer type conversion found.
"crypto/dso/dso_dlfcn.c", line 374.36: 1506-374 (I) Pointer types "char*" and "struct ld_info*" are not compatible.
"crypto/dso/dso_dlfcn.c", line 374.17: 1506-495 (I) Pointer type conversion found.
"crypto/dso/dso_dlfcn.c", line 374.17: 1506-374 (I) Pointer types "struct ld_info*" and "char*" are not compatible.
```
So I stick with the uintptr_t cast here, too. Commit with the other changes follows...

Comment (makr) [reply crypto/dso/dso_dlfcn.c]: @levitte before changing to the uintptr_t cast it actually was a warning:
`"crypto/dso/dso_dlfcn.c", line 370.22: 1506-068 (W) Operation between types "struct ld_info*" and "char*" is not allowed.`

Comment (levitte) [reply crypto/dso/dso_dlfcn.c]: I'll venture a guess that the compiler warns about that mix because that could lead to alignment faults, while it *has* to allow any calculation when `uintptr_t` used.  That would make some kind of sense.  Anyway, this indicates that a `uintptr_t` cast is the most correct way indeed.

File: crypto/dso/dso_dlfcn.c [hunk 9]
```diff
@@ -308,6 +308,77 @@ static int dladdr(void *address, Dl_info *dl)
...
+    do {
+        this_ldi = next_ldi;
+        if ((((uintptr_t)addr >= (uintptr_t)this_ldi->ldinfo_textorg)
+             && ((uintptr_t)addr <
+                 ((uintptr_t)this_ldi->ldinfo_textorg +
+                     this_ldi->ldinfo_textsize)))
+            || (((uintptr_t)addr >= (uintptr_t)this_ldi->ldinfo_dataorg)
+                && ((uintptr_t)addr <
+                    ((uintptr_t)this_ldi->ldinfo_dataorg +
+                        this_ldi->ldinfo_datasize)))) {
```
Comment (dot-asm) [inline crypto/dso/dso_dlfcn.c]: I wonder what would people think of following formatting instead
```
        if ((((uintptr_t)addr >= (uintptr_t)this_ldi->ldinfo_textorg)
             && ((uintptr_t)addr < ((uintptr_t)this_ldi->ldinfo_textorg +
                                    this_ldi->ldinfo_textsize)))
            || (((uintptr_t)addr >= (uintptr_t)this_ldi->ldinfo_dataorg)
                && ((uintptr_t)addr < ((uintptr_t)this_ldi->ldinfo_dataorg +
                                       this_ldi->ldinfo_datasize)))) {
```

Comment (levitte) [reply crypto/dso/dso_dlfcn.c]: Yes, that does look better

Comment (dot-asm) [reply crypto/dso/dso_dlfcn.c]: Another way to improve readability would be to to declare addr as uintptr_t. I mean 
```
static int dladdr(void *ptr, Dl_info *dl)
{
    uintptr_t addr = (uintptr_t)ptr;
    ...
```

Comment (makr) [reply crypto/dso/dso_dlfcn.c]: Yes, that does look far better. Commit follows ...

**Comment by dot-asm** [comment]
I expect there will be a merge request for master that would harmonize this :-)

**Comment by makr** [comment]
PR #5668 was the one for the master, but it is closed already. I nevertheless added the changes to the respective branch. If a new PR is necessary, please just leave a comment here.

**Comment by levitte** [comment]
Merged into 1.1.0.

e4fa7cc3fb9909c6aee411de15a06f918687b8e2 Custome built dladdr() for AIX.

(I hope the final commit message is satisfactory)
